pkg/api/endpoints/util.go

```diff
@@ -83,21 +65,63 @@ func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
 	// Next, build the N-to-M association the API wants.
 	final := []api.EndpointSubset{}
 	for key, ports := range addrSetsToPorts {
-		addrs := []api.EndpointAddress{}
-		for k := range addrSets[key] {
-			addrs = append(addrs, *k)
+		var addrs, notReadyAddrs []api.EndpointAddress
+		for k, ready := range addrSets[key] {
+			if ready {
+				addrs = append(addrs, *k)
+			} else {
+				notReadyAddrs = append(notReadyAddrs, *k)
+			}
 		}
-		final = append(final, api.EndpointSubset{Addresses: addrs, Ports: ports})
+		final = append(final, api.EndpointSubset{Addresses: addrs, NotReadyAddresses: notReadyAddrs, Ports: ports})
 	}
 
 	// Finally, sort it.
 	return SortSubsets(final)
 }
 
-type addressSet map[*api.EndpointAddress]struct{}
+// The sets of hosts must be de-duped, using IP+UID as the key.
+type addressKey struct {
```

Comment:
Mesos.  Their IPs end up being the node's IP, but the pods are different.

On Wed, Sep 16, 2015 at 12:12 PM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/api/endpoints/util.go
> <https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39672451>
> :
>
> >  	}
> >
> >  	// Finally, sort it.
> >  	return SortSubsets(final)
> >  }
> >
> > -type addressSet map[*api.EndpointAddress]struct{}
> > +// The sets of hosts must be de-duped, using IP+UID as the key.
> > +type addressKey struct {
>
> So this is a question for Tim - Tim, why do we want to preserve duplicate
> IPs that come from different target refs? That's what addressKey provides
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13778/files#r39672451>.
>